pubsub/acks_test.go

```diff
@@ -342,3 +344,88 @@ func TestReceiveReturnsErrorFromSendAcks(t *testing.T) {
 		}
 	}
 }
+
+// callbackDriverSub implements driver.Subscription and allows something like
+// monkey patching of both its ReceiveBatch and SendAcks methods.
+type callbackDriverSub struct {
+	driver.Subscription
+	receiveBatch func(context.Context) ([]*driver.Message, error)
+	sendAcks     func(context.Context, []driver.AckID) error
+}
+
+func (s *callbackDriverSub) ReceiveBatch(ctx context.Context, maxMessages int) ([]*driver.Message, error) {
+	return s.receiveBatch(ctx)
+}
+
+func (s *callbackDriverSub) SendAcks(ctx context.Context, acks []driver.AckID) error {
+	return s.sendAcks(ctx, acks)
+}
+
+func (s *callbackDriverSub) IsRetryable(error) bool { return false }
+
+func (s *callbackDriverSub) ErrorCode(error) gcerrors.ErrorCode { return gcerrors.Internal }
+
+// This test detects the root cause of
+// https://github.com/google/go-cloud/issues/1238.
+// If the issue is present, this test times out. The problem was that when
+// there were no messages available from the provider,
+// pubsub.Subscription.Receive would spin trying to get more messages without
+// checking to see if an unrecoverable error had occurred while sending a batch
+// of acks to the provider.
+func TestReceiveReturnsAckErrorOnNoMoreMessages(t *testing.T) {
+	// If SendAcks fails, the error is returned via receive.
+	ctx := context.Background()
+	serr := errors.New("unrecoverable error")
+	receiveHappened := make(chan struct{})
+	ackHappened := make(chan struct{})
+	ds := &callbackDriverSub{
+		// First call to receiveBatch will return a single message.
+		receiveBatch: func(context.Context) ([]*driver.Message, error) {
+			ms := []*driver.Message{{AckID: 1}}
+			return ms, nil
+		},
+		sendAcks: func(context.Context, []driver.AckID) error {
+			ackHappened <- struct{}{}
+			return serr
+		},
+	}
+	sub := pubsub.NewSubscription(ds, nil)
+	defer sub.Shutdown(ctx)
+	m, err := sub.Receive(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+	m.Ack()
+
+	// Second call to receiveBatch will wait for the pull from the
+	// receiveHappened channel below, and return a nil slice of messages.
+	ds.receiveBatch = func(context.Context) ([]*driver.Message, error) {
+		// Subsequent calls to receiveBatch won't wait on receiveHappened,
+		// and will also return nil slices of messages.
+		ds.receiveBatch = func(context.Context) ([]*driver.Message, error) {
+			return nil, nil
+		}
+		receiveHappened <- struct{}{}
+		return nil, nil
+	}
+
+	errc := make(chan error)
+	go func() {
+		_, err := sub.Receive(ctx)
+		errc <- err
+	}()
+
+	// sub.Receive has to get into the loop and then we need to trigger the unrecoverable error.
+	<-receiveHappened
+
+	// Trigger the unrecoverable error.
+	<-ackHappened
+
+	// Wait for sub.Receive to return so we can check the err from sub.Receive against serr.
+	err = <-errc
+
+	ok := gcerrors.Code(err) == gcerrors.Internal && xerrors.Unwrap(err) == serr
```

Comment:
@zombiezen @vangent @shantuo Here is another case. Here Issac makes the good point (in a review comment that somehow got swallowed) that inlining requires a large parenthesized expr:
```
if !(....) 
```
That doesn't bother me, but I see how it could bother someone, so I guess I could go either way on this one.

In general, my position is that these are OK sometimes, but if the resulting line isn't too long or complicated, it's more readable to inline the variable.

Reply:
So in this case, the concern could be addressed in one of two ways:

One way would be to give the individual terms of the conditional distinct names, like:

```go
isInternalError := gcerrors.Code(err) == gcerrors.Internal
isSubscriptionError := xerrors.Unwrap(err) == serr

if !(isInternalError && isSubscriptionError) {
  // ...
}
```

This approach has the benefit of more clearly labeling what the parts of the conditions are trying to test. In fact, this begins to reveal that these are two separate assertions, and would likely benefit from two separate conditionals. So what I'd actually recommend is:

```go
if got := gcerrors.Code(err); got != gcerrors.Internal {
  t.Errorf("error code = %v; want %v", got, gcerrors.Internal)
}
if got := xerrors.Unwrap(err); got == serr {
  t.Errorf("error = %v; want %v", got, serr)
}
```

(Now that I'm staring critically at this, it leaves me wondering whether the test should use `xerrors.Is(err, serr)`. Yay refactoring!)

Another way would been to apply [De Morgan's Law](https://en.wikipedia.org/wiki/De_Morgan%27s_laws) to "factor" the negation to each of the boolean terms, but IMO the above form is more readable anyway.

Reply:
Good idea. Done.